spyder/plugins/ipythonconsole/widgets/shell.py

```diff
@@ -454,18 +458,68 @@ def change_mpl_backend(self, command):
                 self.silent_execute(command)
 
     # ---- Spyder-kernels methods -------------------------------------------
+    def get_editor(self, filename=None):
+        """Get editor for filename and set it as the current editor."""
+        editorstack = self.get_editorstack()
+        if editorstack is None:
+            return None
+
+        if not filename:
+            return editorstack.get_current_editor()
+
+        index = editorstack.has_filename(filename)
+        if index is None:
+            return None
+
+        editor = editorstack.data[index].editor
+        editorstack.set_stack_index(index)
+        return editor
+
     def get_editorstack(self):
         """Get the current editorstack."""
         plugin = self.ipyclient.plugin
         if plugin.main.editor is not None:
             editor = plugin.main.editor
             return editor.get_current_editorstack()
+        raise RuntimeError('No editorstack found.')
 
     def handle_save_files(self):
         """Save the open files."""
+        self.get_editorstack().save()
+
+    def handle_run_cell(self, cell_name, filename=None):
+        """
+        Get cell code from cell name and file name.
+        """
         editorstack = self.get_editorstack()
-        if editorstack is not None:
-            editorstack.save()
+        editorstack.save()
+        editor = self.get_editor(filename)
+
+        if editor is None:
+            raise RuntimeError(
+                "File {} not open in the editor".format(filename))
```

Comment:
Do we really need to raise an error here? Why not simply show a QMessageBox?

Reply:
It makes more sense to display this error in the shell:
if I type `runcell(1, 'foo.py')`, I want an error in the shell if `'foo.py'` is not open.

Reply:
Ok, agreed.